### DISCUSSION TITLE
Newbie needs help running a batch of processes to multiple images.

### DISCUSSION THREAD

**Discussion by dm3d** [ImageMagick/ImageMagick#2281] [Help] [answered]
Hello! I am a newbie and coding is a hobby so bare with me. Sorry :/

I have 1000 images that have information in the R, G, B and so far I have learned `for %f in (*_RMA.tga) ` will seek out each file.

Where my questions begin:

I would like to take the R channel (greyscale), invert it, and save it into a new file as the Alpha channel. Which I believe it would look like:
`convert %f \(-channel R -negate \) \ \ (-swap 0, 3 ) \ %~nf_MS.tga`

Side note: I might need to do the invert in linear color space... but I don't 100% know yet. 

I would also like to take the G channel (greyscale) and save it into the same new file as before but into the R, B, G channels
`convert %f \(-channel G \) \ \ (SOMETHING HERE? I DONT KNOW YET) \ %~nf_MS.tga`

Then I would like to take the B channel and save it as it as an entirely new file 

`convert %f \(-channel B -seperate \) \  %~nf_AO.tga`

Is it possible to do this entire thing in one process? THANKS so much for taking the time to help me!

**Comment by snibgo** [comment]
What version of IM? If v7, I suggest you use "magick", not "convert".

What platform? Your examples seem to be a mix of bash and Windows.

You probably need "-separate". Also, wherever you have "-channel" to select just one channel, you should later  have a corresponding "+channel".

When doing multiple operations on a single input, the construct " ( +clone ... -write abc.tga +delete ) " is useful.

**Reply by dm3d** [reply]
I am using windows. and IM v7, and open to guidance on how I should be scripting it.

I will look into what you have said, feel like I have been moving at about .5 mph in developing this.

**Comment by snibgo** [comment]
Okay. There are many ways to get the results you want. Start by keeping it simple, with just one file, eg toes.png. I write to png outputs so I can easily see them.

```
magick ^
  toes.png +write mpr:INP ^
  -channel R -separate -negate +channel ^
  mpr:INP ^
  +swap ^
  -compose CopyOpacity -composite ^
  -write _01.png +delete ^
  mpr:INP ^
  -channel G -separate +channel ^
  -duplicate 2 ^
  -combine -set colorspace sRGB ^
  -write _02.png +delete ^
  mpr:INP ^
  -channel B -separate +channel ^
  -write _03.png ^
  NULL:
```
Are the images _01.png, _02.png and _03.png as you want them?

If they are, perhaps you can put this command in a loop, as your OP says.

**Reply by dm3d** [reply]
First off, I would like to thank you and offer to buy you a coffee or two for helping me... 

I like your idea for keeping it simple, so in order for that I created a visual to help me try to sudo code.
I am having issues getting a basic for loop to work in windows bash/cmd prompt... its really frustrating me and it appears I am doing everything correctly from the guides I have found online..

In addition, I think the part that is significantly confusing me is the idea of performing multiple unique commands for each channel, and saving them to a different file.  

![image](https://user-images.githubusercontent.com/29493692/87844323-5a4fd200-c879-11ea-9184-019b7705431c.png)

From research I think I need the following commands
https://imagemagick.org/script/command-line-options.php#alpha
https://imagemagick.org/www/script/command-line-options.php#channel
https://imagemagick.org/www/script/command-line-options.php#negate
https://imagemagick.org/script/command-line-options.php#separate


Here is my attempt:

```

## So far this does activate an alpha channel if there isn't one and inverts the red channel, but I cannot seem to copy the red channel into ## the alpha
convert file01.tga -alpha set -channel R -negate +channel -duplicate 0,4 file02.tga

## This is causing the same issue, No success transferring the G channel into the R G and B channel of the new image.
convert file01.tga -channel G -duplicate 0,1,2 file02.tga

##This does take the B channel and separate into the new file, however, I would like this to duplicate into the R, G, B channels 
convert file01.tga -channel B -separate file03.tga

```

I realized I’m still using convert instead of magick, seems to work with both?

**Reply by dm3d** [reply]
I am uploading a file to test with if you'd like. (this forum doesn't allow tga so I converted it to a png)
![file01](https://user-images.githubusercontent.com/29493692/87845059-d13c9900-c880-11ea-852b-18c03e3b5343.png)

**Comment by snibgo** [comment]
As far as I can see, my single command does what you want.

Your three separate commands don't do what you want.

We can take apart my single command to make three commands:
```
magick ^
  %SRC% ^
  ( +clone ^
    -channel R -separate -negate +channel ^
  ) ^
  -compose CopyOpacity -composite ^
  _01a.png

magick ^
  %SRC% ^
  -channel G -separate +channel ^
  -duplicate 2 ^
  -combine -set colorspace sRGB ^
  _02a.png

magick ^
  %SRC% ^
  -channel B -separate +channel ^
  _03a.png
```

**Comment by dm3d** [comment]
I don’t know how to be more clear, there isn’t suppose to be 3 images.  Only 2 images created from the source image. 

Sudo code:

Image 1 to Image 2:
1.) Take Red channel
2.) Invert Red Channel
3.) Duplicate to Alpha
4.) Take Green Channel
5.) Duplicate to R, G, and B channel

= This new image should now have an R,G,B that looks exactly like the G channel from the source image and it should have an alpha channel that looks exactly like the red channel of the source image but inverted

Image 1 to image 3:
1.) Take green channel
2.) Duplicate to R, G, B channel


I’m trying to read up on %SRC%. And mpr:INP But can’t find anything online about what it means, bare with me. Edit: (just figured out this is the file name)

When I read about -duplicate, it refers to duplicating the image?  Is each channel considered an image? So when you have -duplicate 2 Is that duplicating to channel 2? 

I am also not finding anything on CopyOpacity.

**Comment by dm3d** [comment]
Good news, I got a working for loop finally!

```
for %f in (*_RMA.tga) do (
	echo(%f)

)
```

**Comment by snibgo** [comment]
> This new image should now have an R,G,B that looks exactly like the G channel from the source image and it should have an alpha channel that looks exactly like the red channel of the source image but inverted

Ah, okay.
```
%IMG7%magick ^
  %SRC% ^
  ( -clone 0 ^
    -channel G -separate +channel ^
    -duplicate 2 ^
    -combine -set colorspace sRGB ^
  ) ^
  ( -clone 0 ^
    -channel R -separate -negate +channel ^
  ) ^
  -delete 0 ^
  -compose CopyOpacity -composite ^
  _02b.png
```
SRC is an environment variable. Set it to your input filename. Or just use that filename instead of %SRC%.

"mpr:" is an im-memory format for writing and reading images. See https://www.imagemagick.org/script/formats.php

Yes, "-duplicate" duplicates an image. I use this here after "-separate", which makes image(s) from the selected channel(s).

On CopyOpacity, aka copy-opacity, see https://www.imagemagick.org/script/compose.php

**Reply by dm3d** [reply]
Welp, I appreciate the feedback and help. I must just be dumb because this still makes no sense to me.. 

I am confused how you have landed on using `%IMG7%magick` (which doesn't work) instead of just `magick`

Can you confirm I am understanding the arguments/operations correctly?

`-clone 0 ` = means it is duplicating the source image (index 0) which includes the R,G,B channels. **I am not clear about the need to clone.** 
`-channel G -separate +channel` = Can you explain this a bit? I understand it focuses out the Green channel, but what do -separate and +channel contribute too. 
`-duplicate 2` = ... You've full on lost me here, Duplicate says to Specify the count and the image to duplicate by its index. So does duplicate 2 mean it is duplicating the source image 2 times?
`-combine` = What is being combined?
` -delete 0` = The source image, index 0, is being deleted?
`-compose CopyOpacity -composite ` = I am guessing this is the main part that puts the G into RGB of the new image. and Inverted R into the alpha?


This works if I hard code the file name, but as soon as I try to use %SRC% everything breaks, Code example here:

```

for %SRC in (*_RMA.tga) do (

	magick ^
	  %SRC% ^
	  ( -clone 0 ^
		-channel G -separate +channel ^
		-duplicate 2 ^
		-combine -set colorspace sRGB ^
	  ) ^
	  ( -clone 0 ^
		-channel R -separate -negate +channel ^
	  ) ^
	  -delete 0 ^
	  -compose CopyOpacity -composite ^
	  %SRC%_new.png


)

```

Returns: unable to open image %SRC%

**Comment by snibgo** [comment]
My prefix of %IMG7% is the path to magick.exe, which isn't on my system path, so I need an explicit path. You probably don't need this.

My command reads the source image and makes two versions of it. The first version contains the G channel copied into the R and B channels. The second version contains the R channel, negated. Finally it deletes the source image and combines the two versions with "-compose CopyOpacity -composite", which copies the second version into the alpha channel of the first.

"-channel G" is a setting that selects just the G channel for subsequent operations.

"-separate" is an operation that makes a grayscale image from each selected channel of images.

"+channel" is a setting that restores the channel selection to all channels.

"-duplicate 2" makes two copies of the top-most image. It is equivalent to "-duplicate 2,-1".

"-combine" makes one image from all the images in the current list, each image becoming a channel in the new image.

So, as you requested, we have created a RGB image that has all three channels copied from the original G channel. (There are quicker ways of doing the same thing.)

You have figured out loops, with a post using "%f".

**Reply by dm3d** [reply]
Thanks you! I believe I am starting to understand now maybe... Because I finally got the FOR loop to work.

```
for %F in (*_RMA.tga) do magick ^
  "%F" ^
  ( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ) ^
  ( -clone 0 -channel R -separate -negate +channel ) ^
  -delete 0 -compose CopyOpacity -composite "%~nF_MS.tga"

)
```

Ok, so you've been so patient and helpful to get me get this far... what am I missing to tie in the other operation that separates out the B channel into the RGB of the new file, if I try to add that into the same for loop, it breaks. Is that what `mpr:INP` helps with?

**Comment by snibgo** [comment]
Your parentheses don't match. You should have the same number of open ( and close ). I suggest you add an open parenthesis ( between "do" and "magick". You also need to escape every parenthesis in the magick command: ^( and ^).

Now, within the (...) you can add another "magick" command that does whatever you want, such as creating your other result.

Then you can combine both magick commands into one.

**Reply by dm3d** [reply]
OK, I will try and work on getting this to work for the next few days

```
for %F in (*_RMA.tga) do (
  SET _fileName=%~nF
  SET _MStextureName=%_fileName:RMA=MS%
  SET _AOtextureName=%_fileName:RMA=AO%

  magick "%F" ^ 
    ( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ) ^
    ( -clone 0 -channel R -separate -negate +channel ) ^
    -delete 0 -compose CopyOpacity -composite "%_MStextureName%.tga"

)
```

**Reply by dm3d** [reply]
Going on day three, and I feel like I have just tried about every way possible to write the script above like how you described, I am at the point where my IDE editor doesn't think anything is wrong with the script, but it won't run. 

The error that seems to keep coming up

![image](https://user-images.githubusercontent.com/29493692/87958770-2c33e300-ca6f-11ea-8d5a-b125bdd809be.png)




```
for %F in (*_RMA.tga) do (
  SET _fileName=%~nF 
  SET _MStextureName=%_fileName:RMA=MS% 
  SET _AOtextureName=%_fileName:RMA=AO% 

  magick ^( "%F" ^ 
    ( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ) ^
    ( -clone 0 -channel R -separate -negate +channel ) ^
    -delete 0 -compose CopyOpacity -composite "%_MStextureName%.tga" ^)

)

```

![image](https://user-images.githubusercontent.com/29493692/87959650-710c4980-ca70-11ea-8b9e-cc866a766797.png)

**Comment by snibgo** [comment]
> You also need to escape every parenthesis in the magick command: ^( and ^).

You don't seem to have done this. Instead, you have added an extra pair of escaped parentheses.

**Reply by dm3d** [reply]
I mean I could sit here and post every thing I tried but we would be here for awhile.

Yes I did try replacing all of the ( ). With ^( and ^) for the magick command. Same issue

**Reply by dm3d** [reply]
In this example, I felt like it was the closest I have gotten, but still had errors.

```
for %F in (*_RMA.tga) do (
  SET _fileName=%~nF 
  SET _MStextureName=%_fileName:RMA=MS% 
  SET _AOtextureName=%_fileName:RMA=AO% 

  magick "%F" ^
    ^( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ^)
    ^( -clone 0 -channel R -separate -negate +channel ^) 
    -delete 0 -compose CopyOpacity -composite "%_MStextureName%.tga" 

)

```

**Reply by dm3d** [reply]
Seems this works now.... but only for 1 image.



I swear I spent a day with this example. I just ended up rewriting from scratch:
```
for %F in (*_RMA.tga) do (
  SET _fileName=%~nF
  SET _MStextureName=%_fileName:RMA=MS%
  SET _AOtextureName=%_fileName:RMA=AO%
  
  magick "%F" ^
    ^( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
    ^( -clone 0 -channel R -separate -negate +channel ^) ^
    -delete 0 -compose CopyOpacity -composite "%_MStextureName%.tga"

)
```
![image](https://user-images.githubusercontent.com/29493692/87965844-a36e7480-ca79-11ea-91a2-ece240174e14.png)


And most importantly, this script still only focuses on the cmd for one image... and I have yet to do the B channel for the second image.. 

Like so, but still broken
```
for %F in (*_RMA.tga) do (
SET _fileName=%~nF
SET _MStextureName=%_fileName:RMA=MS%
SET _AOtextureName=%_fileName:RMA=AO%
  
  magick "%F" ^
    ^( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
    ^( -clone 0 -channel R -separate -negate +channel ^) ^
    -delete 0 -compose CopyOpacity -composite "%_MStextureName%.tga" ^
	
  magick "%F" ^
	^( -clone 0 -channel B -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
	-delete 0 -compose -composite "%_AOtextureName%.tga"

)


```

**Comment by snibgo** [comment, marked as answer]
At the top of the script, insert the line:
```
setlocal enabledelayedexpansion
```

When you refer to _filename etc within the "for" loop, surround them with !! instead of %%.

```
setlocal enabledelayedexpansion

for %F in (*_RMA.tga) do (
SET _fileName=%~nF
SET _MStextureName=!_fileName:RMA=MS!
SET _AOtextureName=!_fileName:RMA=AO!
  
  magick "%F" ^
    ^( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
    ^( -clone 0 -channel R -separate -negate +channel ^) ^
    -delete 0 -compose CopyOpacity -composite "!_MStextureName!.tga" ^
	
  magick "%F" ^
	^( -clone 0 -channel B -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
	-delete 0 -compose -composite "!_AOtextureName!.tga"

)
```

If this is a BAT script, you need %%F and %%~nF instead of %F and %~nF.

**Reply by dm3d** [reply]
I wish I could send you a venmo/paypal gift for helping me, this meant a lot to me, and thanks again for being so patient.  This script works but i had to save as a .bat file and then drag the .bat file into the cmd prompt.. Learned so much from this I think.

Here was the final code:
```
SETLOCAL ENABLEDELAYEDEXPANSION

FOR /R C:\Users\me\Desktop\IC %%i IN (*_RMA.tga) DO (
set _fileName=%%~ni
set _MStextureName=!_fileName:RMA=MS!
set _AOtextureName=!_fileName:RMA=AO!


  magick "%%i" ^
    ^( -clone 0 -channel G -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
    ^( -clone 0 -channel R -separate -negate +channel ^) ^
    -delete 0 -compose CopyOpacity -composite "!_MStextureName!.tga" ^
	
  magick "%%i" ^
	^( -clone 0 -channel B -separate +channel -duplicate 2 -combine -set colorspace sRGB ^) ^
	-delete 0 "!_AOtextureName!.tga"

)
```
